.github/PULL_REQUEST_TEMPLATE.md

```diff
@@ -1,33 +1,82 @@
-affect other apps (such as the dashboard), make sure the DB columns in the
-various environments are properly populated, coordinate with devops, plan
-migrations in separate steps.
+### General
 
-- [ ] For route changes, make sure GET requests don't change state or result in
-destructive behavior. GET requests should only result in information being
-read, not written.
+- [ ] `make lint` ran locally without errors
 
-- [ ] For encryption changes, make sure it is compatible with data that was
-encrypted with the old code.
+- [ ] Rubocop or Reek offenses were not disabled unless they are false positives.
+If you're not sure, please ask a teammate.
 
-- [ ] For secrets changes, [make sure to update the S3 secrets bucket](https://github.com/18F/identity-private/wiki/Secrets-S3-buckets) with the 
-new configs in **all** environments. 
+- [ ] Tests added for this feature/bug
 
-- [ ] Do not disable Rubocop or Reek offenses unless you are absolutely sure
-they are false positives. If you're not sure how to fix the offense, please
-ask a teammate.
+- [ ] The PR adheres to our [PR guidelines]
+
+- [ ] Related commits have been squashed
+
+- [ ] Method names in classes are in descending order of abstraction.
+See https://youtu.be/0rsilnpU1DU?t=554
+
+### Database
+
+- [ ] Unsafe migrations are implemented over several PRs and over several
+deploys to avoid production errors. The `strong_migrations` gem
+will warn you about unsafe migrations. Please follow the gem's advice.
+
+- [ ] Indexes were added if necessary
+
+- [ ] Verified that the changes don't affect other apps (such as the dashboard)
+
+- [ ] When relevant, an issue has been opened as a reminder to populate the
+necessary DB columns in the various environments right before deploying, taking
+into account the users who might not have interacted with this column yet
+(such as users who have not set a password yet)
+
+- [ ] Migrations against existing tables have been tested against a copy of the
+production database. See #2127 for an example when a migration caused deployment
+issues. In that case, all the migration did was add a new column and an index to
+the Users table, which might seem innocuous.
+
+### Routes
+
+- [ ] GET requests are not vulnerable to CSRF attacks (i.e. they don't change
+state or result in destructive behavior).
+
+### Encryption
+
+- [ ] The changes are compatible with data that was encrypted with the old code.
+
+### Secrets (application.yml)
+
+- [ ] The necessary values have been uploaded to [S3 secrets bucket] with the
+new configs in **all** environments.
+
+### Controllers
+
+- [ ] They adhere to our [controller design guidelines]
+
+- [ ] When adding a new controller that requires the user to be fully
+authenticated, make sure to add `before_action :confirm_two_factor_authenticated`
+as the first callback.
+
+### Service Objects and Form Objects
+
+- [ ] Must only have one public method, usually named `call` or `submit`
```

Comment:
I would differentiate between two kinds of service objects, which might be why we have some of the confusion in that directory. Objects that mediate access to services, such as Twilio, and objects that provide a computational service within the application, such as IdTokenBuilder. The former, which mediate with an outside service, should be able to have multiple methods. The latter can be doable with a single method as long as we name related service objects so we can manage the multiple objects needed for broader services that need multiple actions.

Form objects are interesting because we're getting around the need for multiple methods by returning an object from the single public method that has multiple public methods to represent form state that we don't expect to change on subsequent invocations of the public method. It feels like it's stretching things a bit, but not to a breaking point.

Generally, rules like this should be a rule of thumb that make us stop and consider why we need something and not dogma that holds up delivering for the user.

Reply:
Yes, I can clarify this, and will replace `must` with `should`. There are different types of classes that can be used to perform a specific task or present data. You can call them what you want, but the industry has mostly settled on the ones described here: https://codeclimate.com/blog/7-ways-to-decompose-fat-activerecord-models/

For the most part, these classes will only be responsible for one thing, and will usually only need one method. In the case of Twilio, we have 2 methods: `send_sms` and `place_call`, which makes sense in this context.

Examples of classes that don't need to adhere to this rules include Presenters, View Objects, and Value Objects.